test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,298 @@
+	"sigs.k8s.io/cluster-api/test/e2e/internal/log"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/bootstrap"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ClusterctlUpgradeSpecInput is the input for ClusterctlUpgradeSpec.
+type ClusterctlUpgradeSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+}
+
+// ClusterctlUpgradeSpec implements a test that verifies clusterctl upgrade of a management cluster.
+//
+// NOTE: this test is designed to test v1alpha3 --> v1alpha4 upgrades and v1alpha4 --> v1alpha4.
+func ClusterctlUpgradeSpec(ctx context.Context, inputGetter func() ClusterctlUpgradeSpecInput) {
+	var (
+		specName = "clusterctl-upgrade"
+		input    ClusterctlUpgradeSpecInput
+
+		managementClusterNamespace     *corev1.Namespace
+		managementClusterCancelWatches context.CancelFunc
+		managementClusterResources     *clusterctl.ApplyClusterTemplateAndWaitResult
+		managementClusterProxy         framework.ClusterProxy
+
+		testNamespace     *corev1.Namespace
+		testCancelWatches context.CancelFunc
+		clusterName       string
+	)
+
+	BeforeEach(func() {
+		Expect(ctx).NotTo(BeNil(), "ctx is required for %s spec", specName)
+		input = inputGetter()
+		Expect(input.E2EConfig).ToNot(BeNil(), "Invalid argument. input.E2EConfig can't be nil when calling %s spec", specName)
+		Expect(input.ClusterctlConfigPath).To(BeAnExistingFile(), "Invalid argument. input.ClusterctlConfigPath must be an existing file when calling %s spec", specName)
+		Expect(input.BootstrapClusterProxy).ToNot(BeNil(), "Invalid argument. input.BootstrapClusterProxy can't be nil when calling %s spec", specName)
+		Expect(os.MkdirAll(input.ArtifactFolder, 0755)).To(Succeed(), "Invalid argument. input.ArtifactFolder can't be created for %s spec", specName)
+		Expect(input.E2EConfig.Variables).To(HaveKey(KubernetesVersion))
+
+		// Setup a Namespace where to host objects for this spec and create a watcher for the namespace events.
+		managementClusterNamespace, managementClusterCancelWatches = setupSpecNamespace(context.TODO(), specName, input.BootstrapClusterProxy, input.ArtifactFolder)
+		managementClusterResources = new(clusterctl.ApplyClusterTemplateAndWaitResult)
+	})
+
+	It("Should create a management cluster and then upgrade all the providers", func() {
+		By("Creating a workload cluster to be used as a new management cluster")
+		// NOTE: given that the bootstrap cluster could be shared by several tests, it is not practical to use it for testing clusterctl upgrades.
+		// So we are creating a workload cluster that will be used as a new management cluster where to install older version of providers
+
+		clusterctl.ApplyClusterTemplateAndWait(context.TODO(), clusterctl.ApplyClusterTemplateAndWaitInput{
```

Comment:
we should probably reduce the amount of context.TODOs, either use the ctx which is already there (not sure if that would be correct) or define one at the top, (maybe also context.Background instead of TODO)